cpp/src/arrow/util/borrowed.h

```diff
@@ -0,0 +1,165 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "arrow/util/macros.h"
+#include "arrow/util/mutex.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+namespace internal {
+
+/// A thread safe set of values which can be borrowed (for example by tasks running in a
+/// ThreadPool). Borrowing a value will either lazily construct it or pop from a cache.
+/// Calls to the constructor and destructor are not synchronized.
+template <typename T>
+class BorrowSet : public std::enable_shared_from_this<BorrowSet<T>> {
```

Comment:
Is there precedent in calling this "borrow set"? I would call this a free list, or free list allocator. Or perhaps a "reuse pool".